File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_10.cpp
// *
// **
// ***
// ****
// ***
// **
// *
//----------------------------------
// BRUTE FORCE APPROACH (Two separate loops)
//----------------------------------
#include <iostream>
using namespace std;
int main()
{
    int n = 4;
    // Upper half: 1 to 4 stars
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= i; j++)
        {
            cout << "*";
        }
        cout << endl;
    }
    // Lower half: 3 down to 1 stars
    for (int i = n - 1; i >= 1; i--)
    {
        for (int j = 1; j <= i; j++)
        {
            cout << "*";
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (String constructor)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 4;
    // Upper half
    for (int i = 1; i <= n; i++)
    {
        cout << string(i, '*') << endl;
    }
    // Lower half
    for (int i = n - 1; i >= 1; i--)
    {
        cout << string(i, '*') << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Single loop with formula)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 4;
    for (int i = 1; i <= 2 * n - 1; i++)
    {
        int stars = (i <= n) ? i : 2 * n - i;
        cout << string(stars, '*') << '\n';
    }
    return 0;
}

File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_10.py
# *
# **
# ***
# ****
# ***
# **
# *
#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
n = 4
# Upper half
for i in range(1, n + 1):
    line = ""
    for j in range(i):
        line += "*"  # O(n²) string copying
    print(line)
# Lower half
for i in range(n - 1, 0, -1):
    line = ""
    for j in range(i):
        line += "*"
    print(line)

#----------------------------------
# BETTER APPROACH (String multiplication)
#----------------------------------
n = 4
# Upper half
for i in range(1, n + 1):
    print("*" * i)
# Lower half
for i in range(n - 1, 0, -1):
    print("*" * i)

#----------------------------------
# BEST APPROACH (Single loop with list comprehension)
#----------------------------------
n = 4
pattern = ["*" * i for i in range(1, n + 1)]
pattern.extend(["*" * i for i in range(n - 1, 0, -1)])
print('\n'.join(pattern))